dali/operators/math/expressions/broadcasting.cc

```diff
@@ -13,81 +13,117 @@
+#include <string>
+#include <vector>
 #include "dali/operators/math/expressions/arithmetic_meta.h"
 
 namespace dali {
 
-void CheckNumSamples(span<const TensorListShape<>> shapes) {
-  DALI_ENFORCE(shapes.size() >= 1);
-  int nsamples = shapes[0].num_samples();
+void CheckNumSamples(span<const TensorListShape<>*> shapes) {
+  if (shapes.size() <= 1)
+    return;
+  int nsamples = shapes[0]->num_samples();
   for (int i = 1; i < shapes.size(); i++) {
-    DALI_ENFORCE(nsamples == shapes[i].num_samples());
+    DALI_ENFORCE(nsamples == shapes[i]->num_samples(),
+                 make_string("Number of samples should match. Got shapes with ", nsamples, " and ",
+                             shapes[i]->num_samples(), " respectively."));
   }
 }
 
 template <typename Shape>
-int BroadcastNdimImpl(span<Shape> shapes) {
+int BroadcastNdimImpl(span<const Shape*> shapes) {
   DALI_ENFORCE(shapes.size() >= 1);
-  int ndim = shapes[0].size();
+  int ndim = shapes[0]->sample_dim();
   for (int i = 1; i < shapes.size(); i++) {
-    ndim = std::max(ndim, shapes[i].size());
+    ndim = std::max(ndim, shapes[i]->sample_dim());
   }
   return ndim;
 }
 
-int BroadcastNdim(span<const TensorShape<>> shapes) {
+int BroadcastNdim(span<const TensorShape<>*> shapes) {
   return BroadcastNdimImpl(shapes);
 }
 
-int BroadcastNdim(span<const TensorListShape<>> shapes) {
-  DALI_ENFORCE(shapes.size() >= 1);
-  int ndim = shapes[0].sample_dim();
-  for (int i = 1; i < shapes.size(); i++) {
-    ndim = std::max(ndim, shapes[i].sample_dim());
+int BroadcastNdim(span<const TensorListShape<>*> shapes) {
+  return BroadcastNdimImpl(shapes);
+}
+
+/**
+ * @brief Implementation helper for `BroadcastShape` and `CanBroadcast`
+ *        The function updates val with the d-th dimension of `sh`, or returns false.
+ *        `val` can be updated with a new value if its current value is 1 or equal to the new one.
+ * @tparam ShapeLike e.g. span<const int64_t>, TensorShape<>
+ * @param val extent value to update (in an output shape)
+ * @param sh one of input shape
+ * @param d one of the dimension indices
+ * @return true if val could be updated (broadcastable)
+ * @return false if val could not be updated (non broadcastable)
+ */
+template <typename ShapeLike>
+bool BroadcastShapeImpl(int64_t& val, const ShapeLike& sh, int d) {
+  int ndim = sh.size();
+  auto extent = d < ndim ? sh[ndim - 1 - d] : 1;
+  if (extent > 1 && extent != val) {
+    if (val > 1)
+      return false;
+    val = extent;
   }
-  return ndim;
+  return true;
 }
 
-template <typename Shape>
-void BroadcastShapeImpl(Shape& result, span<const Shape> shapes) {
+std::string PrintShape(const TensorShape<>& shape, int sample_idx, int d) {
+  (void) sample_idx;
+  return make_string(to_string(shape), "(d=", shape.sample_dim() - d - 1, ")");
+}
+
+std::string PrintShape(const TensorListShape<>& shape, int sample_idx, int d) {
+  return make_string(to_string(shape[sample_idx]), " (d=", shape.sample_dim() - d - 1,
```

Comment:
I don't exactly get why do we calculate `d` this way here.

Reply:
This is because we align shapes to the right. I'll give you an example:
Broadcast (10, 10, 3) and (10, 1). We want to compare like this:
```
10, 10, 3
      10, 1
```
Because of this, we iterate from the rightmost dimension to the leftmost, and if d exceeds the available dimension for one shape, we assume 1. I can add a comment to make it clearer